Implementing method to determine the period of some trigonometric functions

kshitij10496 opened the pull request:
This PR is an attempt at computing the period of a simple trigonometric function.
By _simple_, I mean functions of the form : `trig_func(a*x + b)`

The period for such functions is given by : `general_period / Abs(a)`,
where `general_period` is the period of `trig_func(x)`.

Ping @hargup @aktech @asmeurer @Shekharrajak

Commit:
added methods to find the period of trigonometric functions

Commit:
added tests for periodicitiy of trig functions

Shekharrajak commented on sympy/functions/elementary/tests/test_trigonometric.py:
```diff
@@ -1479,3 +1479,20 @@ def test_issue_8653():
+    assert cot(4*x - 6).period() == pi/4
+    assert cos((-3)*x).period() == 2*pi/3
+    assert cos(x*y).period(x) == 2*pi/abs(y)
+    assert sin(3*x*y + 2*pi).period(y) == 2*pi/abs(3*x)
```
[wolfram](https://www.wolframalpha.com/input/?i=period+y%3D+sin(3*x*y+%2B+2*pi)) returns `-2*pi/(3*x)`. Why not same ans ?

kshitij10496 replied:
Because I don't think the period of a function should be negative.
It makes no sense to me to return `-2*pi` as the period for `sin(x)`.

Shekharrajak commented on sympy/functions/elementary/trigonometric.py:
```diff
@@ -14,6 +14,7 @@
 from sympy.functions.elementary.hyperbolic import (acoth, asinh, atanh, cosh,
     coth, HyperbolicFunction, sinh, tanh)
 from sympy.sets.sets import FiniteSet
+from sympy.utilities import filldedent
```
never used.

kshitij10496 replied:
Yeah !
I am a bit confused with whether to use `filldedent` when raising an exception or not.

Ping @aktech @hargup @asmeurer 
I have come across instances where either of them is used.
Do we have a general convention about this issue which I might be overlooking ?

asmeurer replied:
Some folks use it and some folks don't. I would personally use it only if the message is long (and as a side note, putting the string form of an arbitrary expression in an exception message can be an antipattern, as arbitrary expressions can be quite large and this can make raising the exception take much longer than it should). Your two exception messages here are not very long.

kshitij10496 replied:
Thanks.
I will keep this in mind. :+1:

asmeurer commented:
Ideally solveset would get smart enough to compute this sort of thing via `solveset(sin(x + p) - sin(x), p)` (https://github.com/sympy/sympy/issues/10426). I'd like to see a more general `period` function that can compute this for arbitrary functions.

kshitij10496 commented:
This PR attempts only to compute the period of simple trigonometric functions.
I have created this for two reasons:
1. The `solve_decomposition` method is facing certain issues with dealing with infinite critical points and the values of trigonometric functions at the infinities. (`sin(oo)` returns `AccumBounds(-1, 1)`)

2. Extend this property to develop a method to find the periodicity of certain complex trigonometric expressions : `sin(x) + sin(2*x) + sin(3*x)`

As you rightly point out, we need to develop a more general function for determining the period of arbitrary functions (if such a period exists).

Commit:
Method to find the periodicity of general trigonometric functions

Commit:
modified the cases in periodicity to use lcm_list

Commit:
implements method to compute lcm of list of numbers

Commit:
modified behaviour of periodicity for functions with negative coeffs

Commit:
doctest fixes for periodicity and lcm_fraction

Commit:
added tests for periodicity

Commit:
added tests for lcm_fraction

kshitij10496 commented on sympy/calculus/util.py:
```diff
@@ -253,6 +253,79 @@ def elm_domain(expr, intrvl):
+    >>> from sympy.calculus.util import lcm_fraction
+    >>> lcm_fraction([S(1)/2, S(3)/4, S(5)/6])
+    15/2
+    >>> lcm_fraction([2*pi, 3*pi, pi, pi/2])
```
This was one of the cases where `lcm_list` fails.
I would appreciate some discussion on whether to have this functionality or develop a workaround `lcm_list`
so that it works for us.

Ping @asmeurer @jksuom @hargup @aktech

hargup replied:
`lcm_list` calculates the LCM of polynomials, but it should have given the correct result for your case. Please don't create a new function, but rather fix `lcm_list`.

Commit:
periodicity of a general Add object

hargup commented:
@kshitij10496 please update `__init__.py` in `sympy/calculus`. This is a useful function and should be public.

hargup commented on sympy/calculus/util.py:
```diff
@@ -253,6 +253,83 @@ def elm_domain(expr, intrvl):
+    from sympy.functions.elementary.trigonometric import TrigonometricFunction
+
+    orig_f = f
+    f = simplify(orig_f)
```
You should use `trigsimp` here, instead of generic `simplify`.

kshitij10496 replied:
I used `simplify` instead of `trigsimp` because I wanted this functionality to be developed with a general case in mind (non-trigonometric functions). 
I think both give the same output in case of trigonometric functions.

kshitij10496 commented:
Thats cool.

Also, I believe this function can be extended and improved over time to work will general trigonometric functions.

Commit:
added periodicity function to __init__

Shekharrajak commented on sympy/calculus/util.py:
```diff
@@ -253,6 +253,83 @@ def elm_domain(expr, intrvl):
 
+def periodicity(f, symbol):
+    """Returns the period of a given function, if possible.
+
```
Can you write about Parameters, exceptions it raises and for which kind of functions(general form) it is working currently ?

kshitij10496 replied:
Doesn't the 1st line of the docstring give sufficient information about the function ?

Shekharrajak commented on sympy/calculus/util.py:
```diff
@@ -253,6 +253,83 @@ def elm_domain(expr, intrvl):
+    from sympy import lcm, gcd, fraction
+
+    if len(fractions) == 2:
+        a, b = fraction(fractions[0])
```
You can give better name to make it understandable which is numerator and denominator.

kshitij10496 replied:
I think its self-explanatory that we are interested in finding the LCM of `a/b` and `c/d`.
`numer1, numer2` would make the code look counter intuitive.
Won't you agree ?

Shekharrajak replied:
For me `numer`1, ` numer2` is better.

Shekharrajak commented on sympy/calculus/tests/test_util.py:
```diff
@@ -63,6 +62,30 @@ def test_not_empty_in():
+    assert periodicity(sin(4*x) + sin(x)*cos(x), x) == pi
+    assert periodicity(sin(x) - 1, x) == 2*pi
+    assert periodicity(sin(2*x)*cos(2*x) - y, x) == pi/2
+    raises (NotImplementedError, lambda: periodicity(sin(x) - cos(x), x))
```
If it works for `sin(x) + cos(x)` then why not for `sin(x) - cos(x)`. Whats the main problem when coeff is negative?

kshitij10496 replied:
> Whats the main problem when coeff is negative?

The problem is with generalising it.

Shekharrajak commented on sympy/calculus/util.py:
```diff
@@ -253,6 +253,83 @@ def elm_domain(expr, intrvl):
+    if period is not None and orig_f.subs(symbol, symbol + period) == orig_f:
+        return period
+
+    else:
```
```
>>> periodicity(x + sin(x), x)
NotImplementedError: Periods of only certain trigonometric functions can be calculated using this method.
```
Can you add few lines about "certain trigonometric functions" for which it works fine currently.

Shekharrajak commented on sympy/functions/elementary/tests/test_trigonometric.py:
```diff
@@ -1479,3 +1479,20 @@ def test_issue_8653():
+    assert sin(2*x).period() == pi
+    assert cot(4*x - 6).period() == pi/4
+    assert cos((-3)*x).period() == 2*pi/3
+    assert cos(x*y).period(x) == 2*pi/abs(y)
```
```
In [4]: periodicity(cos(x*y),x)

NotImplementedError: Periods of only certain trigonometric functions can be calculated using this method.

In [5]: cos(x*y).period(x)
Out[5]: 
2⋅π
───
│y│

```
both are same right ?

kshitij10496 replied:
The reason this doesn't work with `periodicity` function is due to the check [here](https://github.com/sympy/sympy/pull/11277/files/c72047565116282ed35a42916559bb6d646643fc#diff-a5d2575656fef33ceed5fc4cd005e3c0R303)

```python
In []: f = cos(x*y)
In []: f.subs(x, x + 2*pi/Abs(y))
Out[]: cos(y*(x + 2*pi/Abs(y)))
```
As this cannot be simplified to `cos(x*y)`, I have raised a `NotImplementedError`.

Shekharrajak replied:
Okay then without `Abs` it will work fine.

```
In [1]: periodicity(sec(x) ,x)

NotImplementedError: Periods of only certain trigonometric functions can be calculated using this method.
```
What about these types of cases ?

kshitij10496 replied:
Thanks @Shekharrajak for pointing out the issue of auto simplification with `subs`.
I have added a `check` flag to `periodicity` to mitigate this.
I would love your thoughts on the new implementation.

Commit:
added check flag to periodicity

Commit:
added the case of power functions to periodicity

Commit:
finding periods of multiplication of functions

Commit:
finding the period of composition of function

Commit:
check for argument in standard form in period

Commit:
added helper function to periodicity

Commit:
implemented nlcm function

Commit:
implemented compogen function

Commit:
modified cases for helper and nlcm

Commit:
added tests for periodicity after helper

Commit:
added tests for compogen function

kshitij10496 commented:
I have added few cases based on the suggestion from @aktech that we need not restrict this function to finding the fundamental period of the given expression.

Also, I have developed the following functions for facilitating the implementation:

- `nlcm`(inspired from `ilcm`): aimed at computing the LCM of numbers (both rational and irrational).
- `compogen`: converse of `decompogen`

Reference : [Notes on Periodicity](https://hackmd.io/JwIwLAhmCMBsIFoBMBmAHAdgWArAEzATWAlgQgGM8AGUPJNCEAUyA===)

Ping @sympy/sympy-solvers @asmeurer @smichr 
Let me know how can these be improved with respect to code quality and performance.

Commit:
added docstrings for periodicity and nlcm

kshitij10496 commented:
This PR is ready for review.

hargup commented on sympy/calculus/util.py:
```diff
@@ -253,6 +254,192 @@ def elm_domain(expr, intrvl):
+        return None
+
+
+def nlcm(numbers):
```
Can you create an implementation where `pi` doesn't get a special treatment, also what is the output for `nlcm([2*E, 3*E, E, E/2])`?

jksuom replied:
It seems that this function is concerned with *integral* multiples of numbers. There are many kinds of multiples, depending on the base ring: integral, rational, algebraic, etc.... We can say that `sqrt(2)` is an algebraic multiple of 2, though not a rational multiple; `E` and `pi` are not even algebraic multiples  of each other. If the base ring is not evident from the context, it should probably be indicated explicitly. In this case, the name 'least common integral multiple' could be suitable. (Abbreviated, for example, `lcim` to tell apart from the ordinary `lcm`.)

kshitij10496 replied:
> In this case, the name 'least common integral multiple' could be suitable. (Abbreviated, for example, lcim to tell apart from the ordinary lcm.)

`lcim` sounds good to me too.
 I will update the name accordingly.
Thanks for your input on this !

kshitij10496 replied:
> what is the output for `nlcm([2*E, 3*E, E, E/2])`?

`None` is being returned. 
I will try to generalize this implementation.
In Meeting# 10, @aktech had suggested that an implementation specific to our cause would do for now.
Hence, I had developed this method with the thought of extending it in the future.

hargup commented on sympy/calculus/util.py:
```diff
@@ -253,6 +254,192 @@ def elm_domain(expr, intrvl):
         return _domain
 
 
+def periodicity(f, symbol, check=False):
```
Can you simply call the function `period`?

kshitij10496 replied:
> Can you simply call the function period?

I had named it `periodicity` because:

1. `period` is an instance method of class `TrigonometricFunction`. Having two functions with the same name(and same behaviour) might lead to confusion in the future.

2. I intend to read this function as "Check the function `f` for periodicity in `symbol`. If it is periodic, return the value of the period."

Commit:
generalised nlcm for irrational numbers

Commit:
rename nlcm to lcim

kshitij10496 commented:
I have changed the name of `nlcm` to `lcim` and updated the docs accordingly.
Also, I have tried to implement a generalised method for finding the LCM of irrational numbers.

Ping @sympy/sympy-solvers @asmeurer @jksuom

Shekharrajak commented on sympy/calculus/tests/test_util.py:
```diff
@@ -63,6 +63,52 @@ def test_not_empty_in():
+    assert periodicity(sin(x) - 1, x) == 2*pi
+    assert periodicity(sin(4*x) + sin(x)*cos(x), x) == pi
+    assert periodicity(exp(sin(x)), x) == 2*pi
+    assert periodicity(log(cot(2*x)) - sin(cos(2*x)), x) == pi
```
[function is not periodic](http://www.wolframalpha.com/input/?i=period+y%3D+log(cot(2*x))+-+sin(cos(2*x)))

Shekharrajak replied:
It is periodic in `x`. It is fine.

jksuom replied:
It seems that Wolfram may have a different notion of periodicity of multivalued functions. The function `log(cot(2*x))` has an infinite number branches none of which is periodic. It becomes periodic if `log` is restricted to the principal branch.

Shekharrajak commented on sympy/calculus/tests/test_util.py:
```diff
@@ -63,6 +63,52 @@ def test_not_empty_in():
+    y = Symbol('y')
+
+    assert periodicity(tan(x), x, check=True) == pi
+    assert periodicity(sin(x) + cos(x), x, check=True) == 2*pi
```
```
In [2]: periodicity((sin(x) + cos(x))**2, x)
---------------------------------------------------------------------------
GeneratorsNeeded   
```
check this .

kshitij10496 replied:
Thanks for pointing this out.
I had been meaning to ask this.

What should be the value returned(or exception raised) for a constant function ?
Note : A constant function is period but has no fundamental period.

Ping @jksuom

kshitij10496 replied:
> Perhaps the value returned for a constant could be 0. It is not a period mathematically, but it is different from None and valid periods, and can be easily tested.

I have a small concern here.
We are asking this method (`periodicity`) to return the period of a function with respect to a variable.
And if the concerned function is constant with respect to the variable itself, raising an error would be more suitable, wouldn't it ?

For example:  Currently, we raise a `ValueError` for this function call:
```python
In []: periodicity(sin(y), x)
ValueError: The specified variable is not present in the function.
```
Similarly, would it not be a good idea to raise a `ValueError` stating that the period of a constant function is undefined in the given variable ?

jksuom replied:
I was thinking of what would happen, if we have a sum or a product with a constant argument. We should probably look at each argument separately. If any of the would be aperiodic, then we could return `None`. Otherwise we could return the least common integral multiple of the periods (even if that were not necessarily the fundamental period). If there is constant arguments that return zero, that value could simply be omitted from the argument list of `lcim`. I think that would be easier than preparations for an exception.

kshitij10496 replied:
> I was thinking of what would happen, if we have a sum or a product with a constant argument. We should probably look at each argument separately.

Yes. That's exactly the basis of this implementation. :smile: 

> If there is constant arguments that return zero, that value could simply be omitted from the argument list of lcim.

This sounds good to me. It would definitely be easier to check for `0` rather than catching an exception.

Just to be clear, by `constant arguments` are you referring to actual constants (numbers) or all the arguments that are constant with respect to the given variable ?
For example:  `sin(y)` is independent of `x `. Should we return `0` here as well ?

jksuom replied:
I think that could apply to all arguments that are constant wrt to the given variable, to `sin(y)` as well.

kshitij10496 replied:
> I think that could apply to all arguments that are constant wrt to the given variable, to sin(y) as well.

That's cool ! I will make the necessary changes. :+1:

Commit:
modified the period of a constant function to S.Zero

kshitij10496 commented:
As suggested by @jksuom , the period of a constant function(i.e. the function is constant with respect to the given variable) returns `0`.

kshitij10496 commented:
Ping @sympy/sympy-solvers @asmeurer @jksuom 

This PR is ready for a final review.

jksuom commented on sympy/calculus/util.py:
```diff
@@ -253,6 +254,197 @@ def elm_domain(expr, intrvl):
+        return S.Zero
+
+    if isinstance(f, TrigonometricFunction):
+        arg = f.args[0]
```
Is this used later?

kshitij10496 replied:
> Is this used later?

No, it ain't used later. I will remove this line. 
Thanks

jksuom commented:
Raising a exception each time no period is found may be inconvenient and even misleading from the user's point of view:
```
>>> periodicity(2*x + 3, x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/calculus/util.py", line 350, in periodicity
    return periodicity(g, symbol)
  File "sympy/calculus/util.py", line 342, in periodicity
    period = periodicity(g, symbol)
  File "sympy/calculus/util.py", line 386, in periodicity
    non-periodic.'''))
NotImplementedError: 
Period of the given function             cannot be determined using
this method. The function might be             non-periodic.
```
Perhaps `None` could be returned instead.

kshitij10496 commented:
> Perhaps None could be returned instead.

I think returning `None` is better suited for cases where we can ascertain that the function is aperiodic.
For the sake of correctness of the result, I was a bit a reserved about returning `None` here.

However, most of the test cases that come to my mind are being solved correctly by this function.
So, would it be advisable to return `None` here ?

jksuom commented:
I think that a function that is used for testing periodicity should raise an exception in 'truly exceptional' cases only. It seems that most of the functions that now raise the exception are actually aperiodic.

kshitij10496 commented:
> It seems that most of the functions that now raise the exception are actually aperiodic.

Yes. I think returning `None` should be acceptable here.
I will make these changes. 
Is there anything else that we would like to improve upon here ?

jksuom commented on sympy/calculus/util.py:
```diff
@@ -253,6 +254,197 @@ def elm_domain(expr, intrvl):
+
+
+    Currently, we can find the period of expressions involving trigonometric
+    functons. Note, that the value returned might not be the "fundamental"
```
functons -> functions

kshitij10496 replied:
Thanks for pointing out the typo.
I will be more careful from the next time.

jksuom commented:
Perhaps one thing: It seems that this is about real periods only (at least currently). This could be indicated in the docstring.

kshitij10496 commented:
>  It seems that this is about real periods only (at least currently).

Do you mean that the period returned is always real value ?
Sorry, I don't seem to understand the context of real periods.
Can you be a bit more elaborate ?

Commit:
returning None for undetermined functions in place of raising exception

jksuom commented:
Maybe I did misunderstand your intentions, but it looked like only functions that have real periods were going to be considered. Functions like `exp`, `sinh`, etc., that have imaginary periods were not returning a positive result. So, I was thinking that perhaps this should be pointed out in the docstring, at least in the current version.

kshitij10496 commented:
> Functions like exp, sinh, etc., that have imaginary periods were not returning a positive result.


Both `exp` and `sinh` return `None`, indicating that the function is aperiodic.
Also, `sin(I*x)` returns a `None` value. 
However, all these functions are periodic in nature with complex periods.

I agree that this should be documented. :+1:

Commit:
update docstrings of periodicity

jksuom commented on sympy/calculus/util.py:
```diff
@@ -253,6 +254,198 @@ def elm_domain(expr, intrvl):
+    Notes
+    =====
+
+    Currently, we donot support functions with a complex period.
```
donot -> do not

kshitij10496 commented:
I don't think the build is failing due to the changes that I have committed.
Is this a random build error ?
Should we restart the build ?

jksuom commented:
I have seen this happen before. Random nonsense strings are created and expected to raise a ValueError in Relational. For some reason it does pass. I'll restart the job.

kshitij10496 commented:
Thanks @jksuom

kshitij10496 commented:
Finally the tests pass !
Is this PR good to go now ?

Shekharrajak commented on sympy/calculus/tests/test_util.py:
```diff
@@ -63,6 +63,55 @@ def test_not_empty_in():
+
+    assert periodicity(sin(2*x), x) == pi
+    assert periodicity((-2)*tan(4*x), x) == pi/4
+    assert periodicity(sin(x)**2, x) == 2*pi
```
Ans is `pi`, right ? [wolframalpha ans](http://www.wolframalpha.com/input/?i=period+y%3D+sin(x)**2)

kshitij10496 replied:
The justification of returning `2*pi` is in the docstings. [[Link](https://github.com/sympy/sympy/pull/11277/files#diff-a5d2575656fef33ceed5fc4cd005e3c0R293)]

Shekharrajak replied:
Okay, fine. 

```
In [1]: periodicity(sin(x)**y, y)
Out[1]: 0

In [2]: periodicity(sin(x)**y, k)
Out[2]: 0

In [3]: periodicity(sin(x)**y, x)
Out[3]: 2⋅π
```
These should be `none`? according to [docs](https://github.com/sympy/sympy/pull/11277/files#diff-a5d2575656fef33ceed5fc4cd005e3c0R276)

kshitij10496 replied:
> ```python
> In [1]: periodicity(sin(x)**y, y)
> Out[1]: 0
> ```
Should be `None`. Noted and fixed ! Thanks

> ```python
> In [2]: periodicity(sin(x)**y, k)
> Out[2]: 0
> ```

This is the correct answer since the given function is constant with respect to the variable of concern (`k`).

Shekharrajak replied:
Did you checked this ?

```
In [3]: periodicity(sin(x)**y, x)
Out[3]: 2⋅π
```
It is not periodic, so ans should be none, right ? [wolfram](http://www.wolframalpha.com/input/?i=period+y+%3D+sin(x)**y+in+x)
Add a testcase for this.

kshitij10496 replied:
> ```python
> In [3]: periodicity(sin(x)**y, x)
> Out[3]: 2⋅π
> ```

This is periodic in `x` with a period same as that of `sin(x)`. 
Hence, `2*pi` is the answer.
I think this case is similar to this [testcase](https://github.com/sympy/sympy/pull/11277/files#diff-c637a1ca6aa8a6dd4cd50503f1ccd0fcR87).

kshitij10496 commented:
Ping @jksuom @sympy/sympy-solvers @asmeurer !

jksuom commented:
This looks good to me. It only answers a restricted periodicity question, but it can be a basis for more complete solutions. +1

kshitij10496 commented:
Thanks @jksuom @hargup @Shekharrajak  !
